File: Roadmap/04 - CADENAS DE CARACTERES/python/andresgcastillo.py
# Crear una cadena de caracteres
cadena = "Hola Mundo"

# Acceso a caracteres específicos
print(cadena[0])  # H

# Subcadenas
print(cadena[0:4])  # Hola

# Longitud
print(len(cadena))  # 10

# Concatenación
cadena2 = " Python"
print(cadena + cadena2)  # Hola Mundo Python

# Repetición
print(cadena * 2)  # Hola MundoHola Mundo

# Recorrido
for caracter in cadena:
    print(caracter)

# Conversión a mayúsculas
print(cadena.upper())  # HOLA MUNDO

# Conversión a minúsculas
print(cadena.lower())  # hola mundo

# Reemplazo
print(cadena.replace("Mundo", "Python"))  # Hola Python

# División
print(cadena.split(" "))  # ['Hola', 'Mundo']

# Unión
lista = ['Hola', 'Mundo']
print(" ".join(lista))  # Hola Mundo

# Interpolación
nombre = "Python"
print(f"Hola {nombre}")  # Hola Python

# Verificación
print("Mundo" in cadena)  # True

#Dificultad Extra

def es_palindromo(palabra):
    return palabra == palabra[::-1]

def es_anagrama(palabra1, palabra2):
    return sorted(palabra1) == sorted(palabra2)

def es_isograma(palabra):
    return len(palabra) == len(set(palabra))

# Prueba
palabra1 = "anagram"
palabra2 = "nagaram"

print(f"¿'{palabra1}' es un palíndromo? {es_palindromo(palabra1)}")
print(f"¿'{palabra1}' y '{palabra2}' son anagramas? {es_anagrama(palabra1, palabra2)}")
print(f"¿'{palabra1}' es un isograma? {es_isograma(palabra1)}")

File: Roadmap/01 - OPERADORES Y ESTRUCTURAS DE CONTROL/javascript/Roadmap/04 - CADENAS DE CARACTERES/javascript/andresgcastillo.js
// Crear una cadena de caracteres
let cadena = "Hola Mundo";

// Acceso a caracteres específicos
let primerCaracter = cadena[0]; // "H"

// Subcadenas
let subcadena = cadena.substring(0, 4); // "Hola"

// Longitud
let longitud = cadena.length; // 10

// Concatenación
let cadenaConcatenada = cadena + ", ¿Cómo estás?"; // "Hola Mundo, ¿Cómo estás?"

// Repetición
let cadenaRepetida = cadena.repeat(2); // "Hola MundoHola Mundo"

// Recorrido
for (let i = 0; i < cadena.length; i++) {
  console.log(cadena[i]);
}

// Conversión a mayúsculas
let mayusculas = cadena.toUpperCase(); // "HOLA MUNDO"

// Conversión a minúsculas
let minusculas = cadena.toLowerCase(); // "hola mundo"

// Reemplazo
let cadenaReemplazada = cadena.replace("Mundo", "Andres"); // "Hola Andres"

// División
let cadenaDividida = cadena.split(" "); // ["Hola", "Mundo"]

// Unión
let cadenaUnida = cadenaDividida.join("-"); // "Hola-Mundo"

// Interpolación
let nombre = "Andres";
let cadenaInterpolada = `Hola ${nombre}`; // "Hola Andres"

// Verificación
let contiene = cadena.includes("Mundo"); // true

//Dificultad Extra

function esPalindromo(palabra) {
  let palabraReversa = palabra.split("").reverse().join("");
  return palabra === palabraReversa;
}

function esAnagrama(palabra1, palabra2) {
  let ordenarPalabra = (palabra) => palabra.toLowerCase().split("").sort().join("");
  return ordenarPalabra(palabra1) === ordenarPalabra(palabra2);
}

function esIsograma(palabra) {
  let letras = palabra.toLowerCase().split("");
  let unicas = [...new Set(letras)];
  return letras.length === unicas.length;
}

let palabra1 = "anagram";
let palabra2 = "nagaram";

console.log(`Las palabras son ${palabra1} y ${palabra2}`);

console.log(`¿Son palíndromos?`);
console.log(`Palabra 1: ${esPalindromo(palabra1)}`);
console.log(`Palabra 2: ${esPalindromo(palabra2)}`);

console.log(`¿Son anagramas? ${esAnagrama(palabra1, palabra2)}`);

console.log(`¿Son isogramas?`);
console.log(`Palabra 1: ${esIsograma(palabra1)}`);
console.log(`Palabra 2: ${esIsograma(palabra2)}`);
